File: src/dd_agent/events.py
"""Capture the Claude Agent SDK message stream as a per-stage JSONL event log.

This is "路子一" (the headless SDK path HAPI uses in cli/src/claude/sdk/query.ts):
iterate the SDK message stream and persist each content block (thinking / text /
tool_use / tool_result / result) so the frontend can render live HAPI-style step
cards. The worker previously did `async for _ in query(): pass` and discarded it.

Storage: one JSONL per stage at  {artifacts}/{campaign}/events/{stage}.jsonl.
Each line: {seq, ts, stage, label, type, ...fields}. `label` is the scatter angle
(e.g. "genetic") or "main" for single-session stages, so a stage's parallel angle
sessions interleave into one ordered log distinguished by `label`.

The run thread sets `events_dir_var` (a ContextVar, inherited by the asyncio tasks
the Runner spawns); the worker calls emit(); the API reads via read_stage_events().
"""
from __future__ import annotations

import contextvars
import json
import os
import threading
import time

events_dir_var: contextvars.ContextVar[str | None] = contextvars.ContextVar("dd_events_dir", default=None)

_seq_lock = threading.Lock()
_seq = 0

_MAX_FIELD = 6000  # cap large tool inputs / results so the log stays renderable


def _next_seq() -> int:
    global _seq
    with _seq_lock:
        _seq += 1
        return _seq


def _cap(value):
    """JSON-serializable, size-capped copy of a field (truncate huge strings/dicts)."""
    try:
        s = value if isinstance(value, str) else json.dumps(value, ensure_ascii=False)
    except Exception:
        s = str(value)
    if len(s) > _MAX_FIELD:
        return s[:_MAX_FIELD] + f"… (+{len(s) - _MAX_FIELD} chars)"
    return value if not isinstance(value, str) else value if len(value) <= _MAX_FIELD else s


def emit(stage: str, label: str, type: str, **fields) -> None:
    d = events_dir_var.get()
    if not d:
        return
    os.makedirs(d, exist_ok=True)
    rec = {"seq": _next_seq(), "ts": time.time(), "stage": stage, "label": label, "type": type}
    rec.update({k: _cap(v) for k, v in fields.items()})
    path = os.path.join(d, f"{stage}.jsonl")
    with open(path, "a", encoding="utf-8") as f:
        f.write(json.dumps(rec, ensure_ascii=False) + "\n")


def read_stage_events(artifacts_root: str, campaign: str, stage: str) -> list[dict]:
    path = os.path.join(artifacts_root, campaign, "events", f"{stage}.jsonl")
    if not os.path.exists(path):
        return []
    out: list[dict] = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                out.append(json.loads(line))
            except Exception:
                pass
    out.sort(key=lambda r: r.get("seq", 0))
    return out

File: engine-ts/src/events.ts
// Phase-2 TS port of the per-stage JSONL event log (src/dd_agent/events.py). emit() appends one
// JSON line per event to {artifacts}/{campaign}/events/{stage}.jsonl; readStageEvents() reads them
// back sorted by seq. The Python ContextVar (events_dir_var) → AsyncLocalStorage; the module-global
// seq counter needs no lock (single-threaded event loop). See docs/bun-migration-eval.md Phase 2.
import { AsyncLocalStorage } from "node:async_hooks";
import { appendFileSync, existsSync, mkdirSync, readFileSync } from "node:fs";
import { join } from "node:path";

/** Run a callback with the events dir bound, like the Python run thread setting events_dir_var. */
export const eventsDir = new AsyncLocalStorage<string>();

let _seq = 0;
function nextSeq(): number {
  return ++_seq;
}

const MAX_FIELD = 6000; // cap large tool inputs / results so the log stays renderable

/** JSON-serializable, size-capped copy of a field (truncate huge strings/dicts). */
function cap(value: unknown): unknown {
  let s: string;
  try {
    s = typeof value === "string" ? value : JSON.stringify(value);
  } catch {
    s = String(value);
  }
  if (s.length > MAX_FIELD) return s.slice(0, MAX_FIELD) + `… (+${s.length - MAX_FIELD} chars)`;
  return value;
}

export function emit(stage: string, label: string, type: string, fields: Record<string, unknown> = {}): void {
  const d = eventsDir.getStore();
  if (!d) return;
  mkdirSync(d, { recursive: true });
  const rec: Record<string, unknown> = { seq: nextSeq(), ts: Date.now() / 1000, stage, label, type };
  for (const [k, v] of Object.entries(fields)) rec[k] = cap(v);
  appendFileSync(join(d, `${stage}.jsonl`), JSON.stringify(rec) + "\n", "utf-8");
}

/** Map one Agent SDK stream message (or runAgent's `{__dd_prompt__}` sentinel) to step-card events
 * via emit(), mirroring worker.py _emit_stream. `skipText` drops chatty assistant prose; `withOutcome`
 * includes ResultMessage.result as the card's Outcome (on for search agents, off for scope). Never
 * throws — telemetry must not break the run. */
export function emitStream(stage: string, label: string, msg: any, opts: { skipText?: boolean; withOutcome?: boolean } = {}): void {
  try {
    if (msg && typeof msg === "object" && "__dd_prompt__" in msg) {
      emit(stage, label, "session_start", { prompt: String(msg.__dd_prompt__) });
      return;
    }
    const t = msg?.type;
    if (t === "assistant") {
      for (const b of msg.message?.content ?? []) {
        if (b.type === "thinking") emit(stage, label, "thinking", { text: b.thinking });
        else if (b.type === "text") {
          if (!opts.skipText && b.text && String(b.text).trim()) emit(stage, label, "text", { text: b.text });
        } else if (b.type === "tool_use") emit(stage, label, "tool_use", { tool_id: b.id, name: b.name, input: b.input });
      }
    } else if (t === "user" && Array.isArray(msg.message?.content)) {
      for (const b of msg.message.content) {
        if (b.type === "tool_result") emit(stage, label, "tool_result", { tool_id: b.tool_use_id, content: b.content, is_error: Boolean(b.is_error) });
      }
    } else if (t === "result") {
      const u = msg.usage ?? {};
      const tokens = (u.input_tokens ?? 0) + (u.output_tokens ?? 0);
      emit(stage, label, "result", {
        session_id: msg.session_id ?? null,
        is_error: Boolean(msg.is_error),
        cost: msg.total_cost_usd ?? null,
        num_turns: msg.num_turns ?? null,
        result: opts.withOutcome ? (msg.result ?? null) : null,
        tokens,
      });
    }
  } catch {
    /* telemetry must never break the run */
  }
}

export function readStageEvents(artifactsRoot: string, campaign: string, stage: string): Record<string, unknown>[] {
  const path = join(artifactsRoot, campaign, "events", `${stage}.jsonl`);
  if (!existsSync(path)) return [];
  const out: Record<string, unknown>[] = [];
  for (const line of readFileSync(path, "utf-8").split("\n")) {
    const t = line.trim();
    if (!t) continue;
    try {
      out.push(JSON.parse(t));
    } catch {
      // skip malformed lines, like the Python except: pass
    }
  }
  out.sort((a, b) => ((a.seq as number) ?? 0) - ((b.seq as number) ?? 0));
  return out;
}

/** Test/seam helper: reset the module seq counter so tests are deterministic. */
export function _resetSeq(): void {
  _seq = 0;
}
